Custom `Throwable` not serializable if using `JsonAutoDetect` settings that only detect Fields

### Search before asking

- [x] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

I am trying to serialize and then deserialize an exception. It works in case of a java.lang.IllegalArgumentException, but it does not work for my -- identical -- riskop.MyIllegalArgumentException.


### Version Information

Spring boot 3.5.0, Jackson 2.19.0

### Reproduction

This is my ObjectMapper config:

```java
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE);
        objectMapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
        PolymorphicTypeValidator ptv = BasicPolymorphicTypeValidator.builder()
                        .allowIfSubType(Object.class)
                        .build();
        objectMapper.activateDefaultTyping(ptv, ObjectMapper.DefaultTyping.NON_FINAL, JsonTypeInfo.As.PROPERTY);
```

This is how I serialize / deserialize:

```java
    public void toFromJson(Throwable in) throws Throwable {
        String json1 = objectMapper.writeValueAsString(in);
        Throwable out = objectMapper.readValue(json1, Throwable.class);
        String json2 = objectMapper.writeValueAsString(out);
        throw out;
    }
```

This is working:
```java
        IllegalArgumentException e1 = new IllegalArgumentException();
        try {
            toFromJson.toFromJson(e1);
        }
        catch (IllegalArgumentException e2) {
            // okay, expected
            Assertions.assertEquals(e1.getMessage(), e2.getMessage());
        }
        catch (Throwable t) {
            throw new RuntimeException(t);
        }
```

This is not working, but I think it should:
```java
        MyIllegalArgumentException e1 = new MyIllegalArgumentException();
        try {
            toFromJson.toFromJson(e1);
        }
        catch (MyIllegalArgumentException e2) {
            // okay, expected
            Assertions.assertEquals(e1.getMessage(), e2.getMessage());
        }
        catch (Throwable t) {
            throw new RuntimeException(t);
        }
```

There is a mini project for demonstrating the problem:
https://github.com/riskop/jackson_throwable_to_from_json

There's no difference between MyIllegalArgumentException and the JDK's one:

JDK's IllegalArgumentException:
https://github.com/openjdk/jdk21/blob/master/src/java.base/share/classes/java/lang/IllegalArgumentException.java

My MyIllegalArgumentException
https://github.com/riskop/jackson_throwable_to_from_json/blob/main/src/main/java/riskop/MyIllegalArgumentException.java

just execute 
mvn clean install

stacktrace:
```java
1619 [INFO]  T E S T S
1619 [INFO] -------------------------------------------------------
1939 [INFO] Running riskop.MyTest
11:08:44.999 [main] INFO org.springframework.test.context.support.AnnotationConfigContextLoaderUtils -- Could not detect default configuration classes for test class [riskop.MyTest]: MyTest does not declare any static, non-private, non-final, nested classes annotated with @Configuration.
11:08:45.056 [main] INFO org.springframework.boot.test.context.SpringBootTestContextBootstrapper -- Found @SpringBootConfiguration riskop.Start for test class riskop.MyTest

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/

 :: Spring Boot ::                (v3.5.0)

2025-06-17T11:08:45.273+02:00  INFO 524080 --- [           main] riskop.MyTest                            : Starting MyTest using Java 21.0.1 with PID 524080 (started by riskop in /home/riskop/IdeaProjects/jackson_throwable_to_from_json)
2025-06-17T11:08:45.274+02:00  INFO 524080 --- [           main] riskop.MyTest                            : No active profile set, falling back to 1 default profile: "default"
2025-06-17T11:08:45.625+02:00  INFO 524080 --- [           main] riskop.MyTest                            : Started MyTest in 0.503 seconds (process running for 0.971)
Mockito is currently self-attaching to enable the inline-mock-maker. This will no longer work in future releases of the JDK. Please add Mockito as an agent to your build as described in Mockito's documentation: https://javadoc.io/doc/org.mockito/mockito-core/latest/org.mockito/org/mockito/Mockito.html#0.3
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
WARNING: A Java agent has been loaded dynamically (/home/riskop/.m2/repository/net/bytebuddy/byte-buddy-agent/1.17.5/byte-buddy-agent-1.17.5.jar)
WARNING: If a serviceability tool is in use, please run with -XX:+EnableDynamicAgentLoading to hide this warning
WARNING: If a serviceability tool is not in use, please run with -Djdk.instrument.traceUsage for more information
WARNING: Dynamic loading of agents will be disallowed by default in a future release
3103 [ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.152 s <<< FAILURE! -- in riskop.MyTest
3103 [ERROR] riskop.MyTest.TestMyIAE -- Time elapsed: 0.007 s <<< ERROR!
java.lang.RuntimeException: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Direct self-reference leading to cycle (through reference chain: riskop.MyIllegalArgumentException["cause"])
        at riskop.MyTest.TestMyIAE(MyTest.java:40)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Direct self-reference leading to cycle (through reference chain: riskop.MyIllegalArgumentException["cause"])
        at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
        at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1359)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter._handleSelfReference(BeanPropertyWriter.java:953)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:726)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:760)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeWithType(BeanSerializerBase.java:643)
        at com.fasterxml.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:503)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:342)
        at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4859)
        at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4079)
        at riskop.ToFromJson.toFromJson(ToFromJson.java:18)
        at riskop.MyTest.TestMyIAE(MyTest.java:33)
        ... 3 more

3118 [INFO] 
3118 [INFO] Results:
3118 [INFO] 
```






### Expected behavior

_No response_

### Additional context

background:

I would like to rewrite a http-invoker based communication framework with REST technology. Http-invoker handled exceptions transparently, if the server side caused an exception, the client side just received it. 

I am trying to reproduce that behaviour. In case of errors the server component would serialize the exception to json, pass it to the client ( with rest ) then the client would restore the original exception from json.

--- a/pom.xml
+++ b/pom.xml
@@ -391,7 +391,9 @@
         <jdk>17</jdk>
       </activation>
       <properties>
-        <argLine>--add-opens=java.base/java.util=ALL-UNNAMED</argLine>
+        <argLine>--add-opens=java.base/java.lang=ALL-UNNAMED
+	         --add-opens=java.base/java.util=ALL-UNNAMED
+	</argLine>
       </properties>
       <build>
         <plugins>
@@ -454,7 +456,9 @@
         <jdk>[21,)</jdk>
       </activation>
       <properties>
-        <argLine>--add-opens=java.base/java.util=ALL-UNNAMED</argLine>
+        <argLine>--add-opens=java.base/java.lang=ALL-UNNAMED
+	         --add-opens=java.base/java.util=ALL-UNNAMED
+	</argLine>
       </properties>
       <build>
         <plugins>

--- a/release-notes/VERSION-2.x
+++ b/release-notes/VERSION-2.x
@@ -22,6 +22,9 @@ Project: jackson-databind
 #5192: Record types are broken on Android when using R8
  (reported by @HelloOO7)
  (fix by @pjfanning)
+#5194: Custom `Throwable` not serializable if using `JsonAutoDetect` settings
+ that only detect Fields
+ (reported by @riskop)
 #5197: Add more informative exception for back-references with `record` type
  (fix by Joo-Hyuk K)
 - Generate SBOMs [JSTEP-14]

--- a/src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java
+++ b/src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java
@@ -944,15 +944,27 @@ public class BeanPropertyWriter extends PropertyWriter // which extends
             throws IOException
     {
         if (!ser.usesObjectId()) {
+            boolean writeAsNull = false;
+
             if (prov.isEnabled(SerializationFeature.FAIL_ON_SELF_REFERENCES)) {
                 // 05-Feb-2013, tatu: Usually a problem, but NOT if we are handling
                 // object id; this may be the case for BeanSerializers at least.
                 // 13-Feb-2014, tatu: another possible ok case: custom serializer
                 // (something OTHER than {@link BeanSerializerBase}
                 if (ser instanceof BeanSerializerBase) {
-                    prov.reportBadDefinition(getType(), "Direct self-reference leading to cycle");
+                    // 09-Jul-2025, tatu: [databind#5194] Let's suppress specific case
+                    //   of "cause" for Throwables
+                    if (_isThrowableFieldCause(prov, bean)) {
+                        writeAsNull = true;
+                    } else {
+                        prov.reportBadDefinition(getType(), "Direct self-reference leading to cycle");
+                    }
                 }
-            } else if (prov.isEnabled(SerializationFeature.WRITE_SELF_REFERENCES_AS_NULL)) {
+            } else {
+                writeAsNull = prov.isEnabled(SerializationFeature.WRITE_SELF_REFERENCES_AS_NULL);
+            }
+
+            if (writeAsNull) {
                 if (_nullSerializer != null) {
                     // 23-Oct-2019, tatu: Tricky part -- caller does not specify if it's
                     //   "as property" (in JSON Object) or "as element" (JSON array, via
@@ -970,6 +982,17 @@ public class BeanPropertyWriter extends PropertyWriter // which extends
         return false;
     }
 
+    // Helper method to recognize `Throwable.cause` Field, which has "this" as initialized
+    // value to mean "not set" (`Throwable.getCause()` translates this to `null`).
+    //
+    // @since 2.20
+    private boolean _isThrowableFieldCause(SerializerProvider prov, Object bean) {
+        return bean instanceof Throwable
+                && _member instanceof AnnotatedField
+                && "cause".equals(_name.getValue())
+                ;
+    }
+    
     @Override
     public String toString() {
         StringBuilder sb = new StringBuilder(40);

--- a/src/test/java/com/fasterxml/jackson/databind/ser/jdk/CustomExceptionSer5194Test.java
+++ b/src/test/java/com/fasterxml/jackson/databind/ser/jdk/CustomExceptionSer5194Test.java
@@ -1,0 +1,47 @@
+package com.fasterxml.jackson.databind.ser.jdk;
+
+import org.junit.jupiter.api.Test;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.databind.*;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.databind.testutil.DatabindTestUtil;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+public class CustomExceptionSer5194Test
+    extends DatabindTestUtil
+{
+    static class MyIllegalArgumentException extends RuntimeException {
+        private static final long serialVersionUID = 1L;
+
+        public MyIllegalArgumentException() {
+            super();
+        }
+
+        public MyIllegalArgumentException(String s) {
+            super(s);
+        }
+
+        public MyIllegalArgumentException(String message, Throwable cause) {
+            super(message, cause);
+        }
+
+        public MyIllegalArgumentException(Throwable cause) {
+            super(cause);
+        }
+    }
+
+    // [databind#5194]: failed to serialize custom exception
+    @Test
+    public void test5194() throws Exception {
+        ObjectMapper mapper = JsonMapper.builder()
+                .visibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE)
+                .visibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)
+                .build();
+
+        String json = mapper.writeValueAsString(new MyIllegalArgumentException());
+        assertNotNull(json);
+    }
+}

